flaky test: PGJobContextTest.testExecuteAndCancelSqlCommands

### To reproduce

on mac-other:
```
2024-01-18T15:33:28.9013190Z 2024-01-18T15:33:28.870598Z E i.q.t.TestListener ***** Test Failed *****io.questdb.test.cutlass.pgwire.PGJobContextTest.testExecuteAndCancelSqlCommands[WITH_WAL] duration_ms=284504 : 
2024-01-18T15:33:28.9023750Z java.lang.RuntimeException: Failed on
2024-01-18T15:33:28.9026930Z  ddl: create table tab (b boolean, ts timestamp, sym symbol)
2024-01-18T15:33:28.9034800Z  query: alter table tab add column s1 symbol index, s2 symbol index, s3 symbol index, s4 symbol index, s5 symbol index, s6 symbol index, s7 symbol index, s8 symbol index, s9 symbol index, s10 symbol index, s11 symbol index, s12 symbol index, s13 symbol index, s14 symbol index, s15 symbol index, s16 symbol index, s17 symbol index, s18 symbol index, s19 symbol index
2024-01-18T15:33:28.9044800Z  exception: 
2024-01-18T15:33:28.9045860Z 	at io.questdb.test.cutlass.pgwire.PGJobContextTest.lambda$testExecuteAndCancelSqlCommands$31(PGJobContextTest.java:3155)
2024-01-18T15:33:28.9050560Z 	at io.questdb.test.cutlass.pgwire.PGJobContextTest.lambda$assertWithPgServer$173(PGJobContextTest.java:10056)
2024-01-18T15:33:28.9055880Z 	at io.questdb.test.AbstractCairoTest.lambda$assertMemoryLeak$7(AbstractCairoTest.java:947)
2024-01-18T15:33:28.9063850Z 	at io.questdb.test.tools.TestUtils.assertMemoryLeak(TestUtils.java:618)
2024-01-18T15:33:28.9066610Z 	at io.questdb.test.AbstractCairoTest.assertMemoryLeak(AbstractCairoTest.java:944)
2024-01-18T15:33:28.9072200Z 	at io.questdb.test.AbstractCairoTest.assertMemoryLeak(AbstractCairoTest.java:938)
2024-01-18T15:33:28.9074930Z 	at io.questdb.test.cutlass.pgwire.PGJobContextTest.assertWithPgServer(PGJobContextTest.java:10049)
2024-01-18T15:33:28.9080820Z 	at io.questdb.test.cutlass.pgwire.PGJobContextTest.assertWithPgServer(PGJobContextTest.java:10012)
2024-01-18T15:33:28.9084490Z 	at io.questdb.test.cutlass.pgwire.PGJobContextTest.assertWithPgServer(PGJobContextTest.java:9997)
2024-01-18T15:33:28.9087400Z 	at io.questdb.test.cutlass.pgwire.PGJobContextTest.testExecuteAndCancelSqlCommands(PGJobContextTest.java:2992)
2024-01-18T15:33:28.9092060Z 	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2024-01-18T15:33:28.9096560Z 	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2024-01-18T15:33:28.9100680Z 	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2024-01-18T15:33:28.9136670Z 	at java.lang.reflect.Method.invoke(Method.java:566)
2024-01-18T15:33:28.9177120Z 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
2024-01-18T15:33:28.9201400Z 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2024-01-18T15:33:28.9205430Z 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
2024-01-18T15:33:28.9256770Z 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2024-01-18T15:33:28.9257950Z 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
2024-01-18T15:33:28.9269370Z 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
2024-01-18T15:33:28.9320270Z 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
2024-01-18T15:33:28.9329490Z 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
2024-01-18T15:33:28.9330590Z 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
2024-01-18T15:33:28.9332810Z 	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
2024-01-18T15:33:28.9344770Z 	at java.lang.Thread.run(Thread.java:829)
2024-01-18T15:33:28.9356300Z Caused by: java.lang.RuntimeException: Timed out waiting for command to appear in registry: alter table tab add column s1 symbol index, s2 symbol index, s3 symbol index, s4 symbol index, s5 symbol index, s6 symbol index, s7 symbol index, s8 symbol index, s9 symbol index, s10 symbol index, s11 symbol index, s12 symbol index, s13 symbol index, s14 symbol index, s15 symbol index, s16 symbol index, s17 symbol index, s18 symbol index, s19 symbol index
2024-01-18T15:33:28.9391410Z 	at io.questdb.test.cutlass.pgwire.PGJobContextTest.lambda$testExecuteAndCancelSqlCommands$31(PGJobContextTest.java:3118)
2024-01-18T15:33:28.9447650Z 	...24 more
```


```
2024-01-18T15:28:44.3684590Z 2024-01-18T15:28:44.366518Z I i.q.t.c.p.PGJobContextTest random seeds: 1572992513180L, 1705591724366L
[...]
```

full log: https://gist.github.com/jerrinot/74657b3fde6ae5e70d71a28745112b66

### QuestDB version:

7.3.10-SNAPSHOT

### OS, in case of Docker specify Docker and the Host OS:

mac-other

### File System, in case of Docker specify Host File System:

irrelevant, probably

### Full Name:

Jaromir Hamala

### Affiliation:

QuestDB

### Have you followed Linux, MacOs kernel configuration steps to increase Maximum open files and Maximum virtual memory areas limit?

- [X] Yes, I have

### Additional context

_No response_

--- a/core/src/main/java/io/questdb/griffin/QueryRegistry.java
+++ b/core/src/main/java/io/questdb/griffin/QueryRegistry.java
@@ -34,6 +34,7 @@ import io.questdb.std.*;
 import io.questdb.std.datetime.microtime.MicrosecondClock;
 import io.questdb.std.str.StringSink;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.TestOnly;
 
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicLong;
@@ -48,6 +49,8 @@ public class QueryRegistry {
     private final AtomicLong idSeq = new AtomicLong();
     private final ConcurrentLongHashMap<Entry> registry = new ConcurrentLongHashMap<>();
     private final ThreadLocal<WeakMutableObjectPool<Entry>> tlQueryPool;
+
+    private volatile Listener listener;
 
     public QueryRegistry(CairoConfiguration configuration) {
         this.clock = configuration.getMicrosecondClock();
@@ -136,8 +139,18 @@ public class QueryRegistry {
         e.principal = executionContext.getSecurityContext().getPrincipal();
         registry.put(queryId, e);
 
+        Listener listener = this.listener;
+        if (listener != null) {
+            listener.onRegister(query, queryId);
+        }
+
         executionContext.setCancelledFlag(e.cancelled);
         return queryId;
+    }
+
+    @TestOnly
+    public void setListener(Listener listener) {
+        this.listener = listener;
     }
 
     /**
@@ -161,6 +174,10 @@ public class QueryRegistry {
         }
 
         executionContext.setCancelledFlag(null);
+    }
+
+    public interface Listener {
+        void onRegister(CharSequence query, long queryId);
     }
 
     public static class Entry implements Mutable {

--- a/core/src/test/java/io/questdb/test/cutlass/http/IODispatcherTest.java
+++ b/core/src/test/java/io/questdb/test/cutlass/http/IODispatcherTest.java
@@ -36,10 +36,7 @@ import io.questdb.cutlass.http.processors.HealthCheckProcessor;
 import io.questdb.cutlass.http.processors.JsonQueryProcessor;
 import io.questdb.cutlass.http.processors.StaticContentProcessor;
 import io.questdb.cutlass.http.processors.TextImportProcessor;
-import io.questdb.griffin.SqlCompiler;
-import io.questdb.griffin.SqlException;
-import io.questdb.griffin.SqlExecutionContext;
-import io.questdb.griffin.SqlExecutionContextImpl;
+import io.questdb.griffin.*;
 import io.questdb.griffin.engine.functions.rnd.SharedRandom;
 import io.questdb.griffin.engine.functions.test.TestDataUnavailableFunctionFactory;
 import io.questdb.griffin.engine.functions.test.TestLatchedCounterFunctionFactory;
@@ -8350,13 +8347,7 @@ public class IODispatcherTest extends AbstractTest {
         String update2 = "update tab set b=sleep(120000)";
         String updateWithJoin1 = "update tab t1 set b=true from tab t2 where sleep(120000) and t1.b = t2.b";
         String updateWithJoin2 = "update tab t1 set b=sleep(120000) from tab t2 where t1.b = t2.b";
-
-        // add many symbols to slow down operation enough so that other thread can detect it in registry and cancel it
-        StringBuilder addColumnsTmp = new StringBuilder("alter table tab add column s1 symbol index");
-        for (int i = 2; i < 30; i++) {
-            addColumnsTmp.append(", s").append(i).append(" symbol index");
-        }
-        final String addColumns = addColumnsTmp.toString();
+        String addColumns = "alter table tab add column s1 symbol index";
 
         final ObjList<String> commands;
         if ("/query".equals(url)) {
@@ -8402,7 +8393,8 @@ public class IODispatcherTest extends AbstractTest {
                                 .withServerKeepAlive(true)
                 )
                 .run((engine) -> {
-                    //testHttpClient.setKeepConnection(true);
+                    DelayedListener registryListener = new DelayedListener();
+                    engine.getQueryRegistry().setListener(registryListener);
 
                     try (SqlExecutionContext executionContext = TestUtils.createSqlExecutionCtx(engine)) {
                         for (int i = 0, n = ddls.size(); i < n; i++) {
@@ -8460,6 +8452,9 @@ public class IODispatcherTest extends AbstractTest {
                                     stopped.setCount(isWal ? 2 : 1);
                                     queryError.set(null);
 
+                                    registryListener.queryText = command;
+                                    registryListener.queryFound.setCount(1);
+
                                     new Thread(() -> {
                                         String response;
                                         try (TestHttpClient testHttpClient = new TestHttpClient()) {
@@ -8506,34 +8501,36 @@ public class IODispatcherTest extends AbstractTest {
                                     long start = System.currentTimeMillis();
 
                                     //wait until query appears in registry and get query id
-                                    while (true) {
-                                        Os.sleep(1);
-                                        //noinspection Convert2Diamond
-                                        testHttpClient.assertGetRegexp(
-                                                "/query",
-                                                ".*dataset.*",
-                                                "select query_id from query_activity() where query = '" + command.replace("'", "''") + "'",
-                                                null, null, null,
-                                                new CharSequenceObjHashMap<String>() {{
-                                                    put("nm", "true");
-                                                }},
-                                                "200"
-                                        );
-                                        String response = testHttpClient.getSink().toString();
-                                        int startIdx = response.indexOf("\"dataset\":[[");
-                                        if (startIdx > -1) {
-                                            startIdx += "\"dataset\":[[".length();
-                                            int endIdx = response.indexOf("]]", startIdx);
-                                            queryId = Numbers.parseLong(response, startIdx, endIdx);
-                                            break;
+                                    try {
+                                        while (true) {
+                                            Os.sleep(1);
+                                            testHttpClient.assertGetRegexp(
+                                                    "/query",
+                                                    ".*dataset.*",
+                                                    "select query_id from query_activity() where query = '" + command.replace("'", "''") + "'",
+                                                    null, null, null,
+                                                    new CharSequenceObjHashMap<String>() {{
+                                                        put("nm", "true");
+                                                    }},
+                                                    "200"
+                                            );
+                                            String response = testHttpClient.getSink().toString();
+                                            int startIdx = response.indexOf("\"dataset\":[[");
+                                            if (startIdx > -1) {
+                                                startIdx += "\"dataset\":[[".length();
+                                                int endIdx = response.indexOf("]]", startIdx);
+                                                queryId = Numbers.parseLong(response, startIdx, endIdx);
+                                                break;
+                                            }
+                                            if (System.currentTimeMillis() - start > TIMEOUT) {
+                                                throw new RuntimeException("Timed out waiting for command to appear in registry: " + command);
+                                            }
+                                            if (queryError.get() != null) {
+                                                throw new RuntimeException("Query to cancel failed!", queryError.get());
+                                            }
                                         }
-
-                                        if (System.currentTimeMillis() - start > TIMEOUT) {
-                                            throw new RuntimeException("Timed out waiting for command to appear in registry: " + command);
-                                        }
-                                        if (queryError.get() != null) {
-                                            throw new RuntimeException("Query to cancel failed!", queryError.get());
-                                        }
+                                    } finally {
+                                        registryListener.queryFound.countDown();
                                     }
 
                                     testHttpClient.assertGetRegexp(
@@ -8582,6 +8579,8 @@ public class IODispatcherTest extends AbstractTest {
                                 }
                             }
                         }
+                    } finally {
+                        engine.getQueryRegistry().setListener(null);
                     }
                 });
     }
@@ -9049,6 +9048,23 @@ public class IODispatcherTest extends AbstractTest {
         @Override
         public int length() {
             return len;
+        }
+    }
+
+    private static class DelayedListener implements QueryRegistry.Listener {
+        private SOCountDownLatch queryFound = new SOCountDownLatch(1);
+        private volatile CharSequence queryText;
+
+        @Override
+        public void onRegister(CharSequence query, long queryId) {
+            if (queryText == null) {
+                return;
+            }
+
+            if (Chars.equalsNc(queryText, query)) {
+                queryFound.await();
+                queryText = null;
+            }
         }
     }
 

--- a/core/src/test/java/io/questdb/test/cutlass/pgwire/PGJobContextTest.java
+++ b/core/src/test/java/io/questdb/test/cutlass/pgwire/PGJobContextTest.java
@@ -34,6 +34,7 @@ import io.questdb.cutlass.pgwire.CircuitBreakerRegistry;
 import io.questdb.cutlass.pgwire.PGWireConfiguration;
 import io.questdb.cutlass.pgwire.PGWireServer;
 import io.questdb.griffin.QueryFutureUpdateListener;
+import io.questdb.griffin.QueryRegistry;
 import io.questdb.griffin.SqlException;
 import io.questdb.griffin.SqlExecutionContextImpl;
 import io.questdb.griffin.engine.functions.test.TestDataUnavailableFunctionFactory;
@@ -2961,14 +2962,7 @@ if __name__ == "__main__":
         String update2 = "update tab set b=sleep(120000)";
         String updateWithJoin1 = "update tab t1 set b=true from tab t2 where sleep(120000) and t1.b = t2.b";
         String updateWithJoin2 = "update tab t1 set b=sleep(120000) from tab t2 where t1.b = t2.b";
-
-        // add many symbols to slow down operation enough so that other thread can detect it in registry and cancel it
-        StringSink addColumnsTmp = new StringSink();
-        addColumnsTmp.put("alter table tab add column s1 symbol index");
-        for (int i = 2; i < 20; i++) {
-            addColumnsTmp.put(", s").put(i).put(" symbol index");
-        }
-        final String addColumns = addColumnsTmp.toString();
+        String addColumns = "alter table tab add column s1 symbol index";
 
         ObjList<String> commands = new ObjList<>(
                 createAsSelect,
@@ -2989,179 +2983,190 @@ if __name__ == "__main__":
                 addColumns
         );
 
-        assertWithPgServer(CONN_AWARE_ALL, (connection, binary, mode, port) -> {
-            SOCountDownLatch started = new SOCountDownLatch(1);
-            SOCountDownLatch stopped = new SOCountDownLatch(1);
-            AtomicReference<Exception> queryError = new AtomicReference<>();
+        try {
+            DelayedListener registryListener = new DelayedListener();
+            engine.getQueryRegistry().setListener(registryListener);
 
-            for (int i = 0, n = ddls.size(); i < n; i++) {
-                final String ddl = ddls.getQuick(i);
-                boolean isWal = ddl.equals(walTable);
+            assertWithPgServer(CONN_AWARE_ALL, (connection, binary, mode, port) -> {
+                SOCountDownLatch started = new SOCountDownLatch(1);
+                SOCountDownLatch stopped = new SOCountDownLatch(1);
+                AtomicReference<Exception> queryError = new AtomicReference<>();
 
-                drop("drop table if exists tab");
-                ddl(ddl);
-                insert("insert into tab select true, (86400000000*x)::timestamp, null from long_sequence(1000)");
-                drop("drop table if exists new_tab");
-                if (isWal) {
-                    drainWalQueue();
-                }
+                for (int i = 0, n = ddls.size(); i < n; i++) {
+                    final String ddl = ddls.getQuick(i);
+                    boolean isWal = ddl.equals(walTable);
 
-                for (int j = 0, k = commands.size(); j < k; j++) {
-                    final String command = commands.getQuick(j);
-
-                    // skip pending wal changes in case wal table has been suspended by earlier command
+                    drop("drop table if exists tab");
+                    ddl(ddl);
+                    insert("insert into tab select true, (86400000000*x)::timestamp, null from long_sequence(1000)");
+                    drop("drop table if exists new_tab");
                     if (isWal) {
-                        try (RecordCursorFactory factory = select("select suspended, writerTxn, sequencerTxn from wal_tables() where name = 'tab'")) {
-                            boolean suspended;
-                            long sequencerTxn;
+                        drainWalQueue();
+                    }
 
-                            try (RecordCursor cursor = factory.getCursor(sqlExecutionContext)) {
-                                cursor.hasNext();
-                                Record record = cursor.getRecord();
-                                suspended = record.getBool(0);
-                                sequencerTxn = record.getLong(2) + 1;
-                            }
+                    for (int j = 0, k = commands.size(); j < k; j++) {
+                        final String command = commands.getQuick(j);
 
-                            if (suspended) {
-                                ddl("alter table tab resume wal from txn " + sequencerTxn);
+                        // skip pending wal changes in case wal table has been suspended by earlier command
+                        if (isWal) {
+                            try (RecordCursorFactory factory = select("select suspended, writerTxn, sequencerTxn from wal_tables() where name = 'tab'")) {
+                                boolean suspended;
+                                long sequencerTxn;
 
                                 try (RecordCursor cursor = factory.getCursor(sqlExecutionContext)) {
                                     cursor.hasNext();
                                     Record record = cursor.getRecord();
-                                    suspended = record.getBool(1);
-                                    Assert.assertFalse(suspended);
+                                    suspended = record.getBool(0);
+                                    sequencerTxn = record.getLong(2) + 1;
+                                }
+
+                                if (suspended) {
+                                    ddl("alter table tab resume wal from txn " + sequencerTxn);
+
+                                    try (RecordCursor cursor = factory.getCursor(sqlExecutionContext)) {
+                                        cursor.hasNext();
+                                        Record record = cursor.getRecord();
+                                        suspended = record.getBool(1);
+                                        Assert.assertFalse(suspended);
+                                    }
                                 }
                             }
-                        }
-                        drainWalQueue();
-                    }
-
-                    // statements containing multiple transactions, such as 'alter table add column col1, col2' are currently not supported for WAL tables
-                    // UPDATE statements with join are not supported yet for WAL tables
-                    if ((isWal && (command.equals(updateWithJoin1) || command.equals(updateWithJoin2) || command.equals(addColumns)))) {
-                        continue;
-                    }
-
-                    try {
-                        if (isWal) {
                             drainWalQueue();
                         }
 
-                        started.setCount(isWal ? 2 : 1);
-                        stopped.setCount(isWal ? 2 : 1);
-                        queryError.set(null);
+                        // statements containing multiple transactions, such as 'alter table add column col1, col2' are currently not supported for WAL tables
+                        // UPDATE statements with join are not supported yet for WAL tables
+                        if ((isWal && (command.equals(updateWithJoin1) || command.equals(updateWithJoin2) || command.equals(addColumns)))) {
+                            continue;
+                        }
 
-                        new Thread(() -> {
-                            try {
-                                try (Connection conn = getConnection(mode, port, binary, 5)) {
-                                    started.countDown();
+                        try {
+                            if (isWal) {
+                                drainWalQueue();
+                            }
 
-                                    try (PreparedStatement stmt = conn.prepareStatement(command)) {
-                                        if (command.startsWith("select")) {
-                                            try (ResultSet result = stmt.executeQuery()) {
-                                                while (result.next()) {
-                                                    //ignore
+                            started.setCount(isWal ? 2 : 1);
+                            stopped.setCount(isWal ? 2 : 1);
+                            queryError.set(null);
+                            registryListener.queryFound.setCount(1);
+                            registryListener.queryText = command;
+
+                            new Thread(() -> {
+                                try {
+                                    try (Connection conn = getConnection(mode, port, binary, 5)) {
+                                        started.countDown();
+
+                                        try (PreparedStatement stmt = conn.prepareStatement(command)) {
+                                            if (command.startsWith("select")) {
+                                                try (ResultSet result = stmt.executeQuery()) {
+                                                    while (result.next()) {
+                                                        //ignore
+                                                    }
                                                 }
+                                            } else {
+                                                stmt.executeUpdate();
                                             }
-                                        } else {
-                                            stmt.executeUpdate();
                                         }
                                     }
-                                }
-                            } catch (SQLException e) {
-                                // ignore errors showing that statement has been cancelled
-                                if (!Chars.contains(e.getMessage(), "Could not create table") &&
-                                        !Chars.contains(e.getMessage(), "cancelled by user")) {
-                                    queryError.set(e);
-                                }
-                            } catch (Exception e) {
-                                queryError.set(e);
-                            } finally {
-                                stopped.countDown();
-                            }
-                        }, "command_thread").start();
-
-                        if (isWal) {
-                            Thread walJob = new Thread(() -> {
-                                started.countDown();
-
-                                try (ApplyWal2TableJob walApplyJob = createWalApplyJob()) {
-                                    while (queryError.get() == null) {
-                                        drainWalQueue(walApplyJob);
+                                } catch (SQLException e) {
+                                    // ignore errors showing that statement has been cancelled
+                                    if (!Chars.contains(e.getMessage(), "Could not create table") &&
+                                            !Chars.contains(e.getMessage(), "cancelled by user")) {
+                                        queryError.set(e);
                                     }
+                                } catch (Exception e) {
+                                    queryError.set(e);
                                 } finally {
-                                    // release native path memory used by the job
-                                    Path.PATH.get().close();
                                     stopped.countDown();
                                 }
-                            }, "wal_job");
-                            walJob.start();
-                        }
+                            }, "command_thread").start();
 
-                        started.await();
+                            if (isWal) {
+                                Thread walJob = new Thread(() -> {
+                                    started.countDown();
 
-                        long queryId;
-                        long start = System.currentTimeMillis();
+                                    try (ApplyWal2TableJob walApplyJob = createWalApplyJob()) {
+                                        while (queryError.get() == null) {
+                                            drainWalQueue(walApplyJob);
+                                        }
+                                    } finally {
+                                        // release native path memory used by the job
+                                        Path.PATH.get().close();
+                                        stopped.countDown();
+                                    }
+                                }, "wal_job");
+                                walJob.start();
+                            }
 
-                        try (PreparedStatement stmt = connection.prepareStatement("select query_id from query_activity() where query = ?")) {
-                            stmt.setString(1, command);
-                            while (true) {
-                                Os.sleep(1);
-                                try (ResultSet result = stmt.executeQuery()) {
-                                    if (result.next()) {
-                                        queryId = result.getLong(1);
-                                        break;
+                            started.await();
+
+                            long queryId;
+                            long start = System.currentTimeMillis();
+
+                            try (PreparedStatement stmt = connection.prepareStatement("select query_id from query_activity() where query = ?")) {
+                                stmt.setString(1, command);
+                                while (true) {
+                                    Os.sleep(1);
+                                    try (ResultSet result = stmt.executeQuery()) {
+                                        if (result.next()) {
+                                            queryId = result.getLong(1);
+                                            break;
+                                        }
+                                    }
+
+                                    if (System.currentTimeMillis() - start > TIMEOUT) {
+                                        throw new RuntimeException("Timed out waiting for command to appear in registry: " + command);
+                                    }
+                                    if (queryError.get() != null) {
+                                        throw new RuntimeException("Query to cancel failed!", queryError.get());
                                     }
                                 }
-
-                                if (System.currentTimeMillis() - start > TIMEOUT) {
-                                    throw new RuntimeException("Timed out waiting for command to appear in registry: " + command);
-                                }
-                                if (queryError.get() != null) {
-                                    throw new RuntimeException("Query to cancel failed!", queryError.get());
-                                }
+                            } finally {
+                                registryListener.queryFound.countDown();
                             }
-                        }
 
-                        try (PreparedStatement stmt = connection.prepareStatement("cancel query " + queryId)) {
-                            stmt.executeUpdate();
-                        }
+                            try (PreparedStatement stmt = connection.prepareStatement("cancel query " + queryId)) {
+                                stmt.executeUpdate();
+                            }
 
-                        start = System.currentTimeMillis();
+                            start = System.currentTimeMillis();
 
-                        try (PreparedStatement stmt = connection.prepareStatement("select * from query_activity() where query_id = ?")) {
-                            stmt.setLong(1, queryId);
-                            while (true) {
-                                Os.sleep(1);
-                                try (ResultSet result = stmt.executeQuery()) {
-                                    if (!result.next()) {
-                                        break;
+                            try (PreparedStatement stmt = connection.prepareStatement("select * from query_activity() where query_id = ?")) {
+                                stmt.setLong(1, queryId);
+                                while (true) {
+                                    Os.sleep(1);
+                                    try (ResultSet result = stmt.executeQuery()) {
+                                        if (!result.next()) {
+                                            break;
+                                        }
+                                    }
+                                    if (System.currentTimeMillis() - start > TIMEOUT) {
+                                        throw new RuntimeException("Timed out waiting for command to stop: " + command);
                                     }
                                 }
-                                if (System.currentTimeMillis() - start > TIMEOUT) {
-                                    throw new RuntimeException("Timed out waiting for command to stop: " + command);
+                            }
+
+                            // run simple query to test that previous query cancellatio doesn't 'spill into' other queries
+                            try (PreparedStatement stmt = connection.prepareStatement("select sleep(1)")) {
+                                try (ResultSet result = stmt.executeQuery()) {
+                                    result.next();
+                                    Assert.assertTrue(result.getBoolean(1));
                                 }
                             }
+                        } catch (Throwable t) {
+                            throw new RuntimeException("Failed on\n ddl: " + ddl +
+                                    "\n query: " + command +
+                                    "\n exception: ", t);
+                        } finally {
+                            queryError.set(new Exception());//stop wal thread
+                            stopped.await();
                         }
-
-                        // run simple query to test that previous query cancellatio doesn't 'spill into' other queries
-                        try (PreparedStatement stmt = connection.prepareStatement("select sleep(1)")) {
-                            try (ResultSet result = stmt.executeQuery()) {
-                                result.next();
-                                Assert.assertTrue(result.getBoolean(1));
-                            }
-                        }
-                    } catch (Throwable t) {
-                        throw new RuntimeException("Failed on\n ddl: " + ddl +
-                                "\n query: " + command +
-                                "\n exception: ", t);
-                    } finally {
-                        queryError.set(new Exception());//stop wal thread
-                        stopped.await();
                     }
                 }
-            }
-        });
+            });
+        } finally {
+            engine.getQueryRegistry().setListener(null);
+        }
     }
 
     @Test
@@ -9289,10 +9294,6 @@ create table tab as (
         });
     }
 
-    //
-    // Tests for ResultSet.setFetchSize().
-    //
-
     @Test
     public void testUpdateAfterDroppingColumnNotUsedByTheUpdate() throws Exception {
         assertMemoryLeak(() -> {
@@ -9330,6 +9331,10 @@ create table tab as (
         });
     }
 
+    //
+    // Tests for ResultSet.setFetchSize().
+    //
+
     @Test
     public void testUpdateAfterDroppingColumnUsedByTheUpdate() throws Exception {
         assertMemoryLeak(() -> {
@@ -9364,10 +9369,6 @@ create table tab as (
             }
         });
     }
-
-    //
-    // Tests for ResultSet.setFetchSize().
-    //
 
     @Test
     public void testUpdateAsync() throws Exception {
@@ -9409,6 +9410,10 @@ create table tab as (
                         "9,3.0,2020-06-01 00:00:12.0,null\n"
         );
     }
+
+    //
+    // Tests for ResultSet.setFetchSize().
+    //
 
     @Test
     public void testUpdateBatch() throws Exception {
@@ -11621,6 +11626,23 @@ create table tab as (
     @FunctionalInterface
     interface ResultProducer {
         void produce(String[] paramVals, boolean[] isBindVals, String[] bindVals, Utf16Sink output);
+    }
+
+    private static class DelayedListener implements QueryRegistry.Listener {
+        private SOCountDownLatch queryFound = new SOCountDownLatch(1);
+        private volatile CharSequence queryText;
+
+        @Override
+        public void onRegister(CharSequence query, long queryId) {
+            if (queryText == null) {
+                return;
+            }
+
+            if (Chars.equalsNc(queryText, query)) {
+                queryFound.await();
+                queryText = null;
+            }
+        }
     }
 
     private static class DelayingNetworkFacade extends NetworkFacadeImpl {

